### ISSUE TITLE
Can't infer inline variadic type concatenation

### ISSUE THREAD

**Issue by ehllie** [microsoft/pyright#4451]
**Describe the bug**
Types of multiple arguments do not get concatenated into the type of the variadic argument of the called function when one of the positional arguments includes an unpack expression.

**To Reproduce**
When calling a function that accepts a variadic argument with multiple arguments, one of which includes the `*` unpack expression, pyright can't properly concatenate them into a single variadic argument. However, when first constructing a tuple from those arguments, and then unpacking said tuple, the type concatenation is performed correctly.

**Expected behavior**
Type of the variadic argument should be concatenated without having to go through an intermediary tuple.

**Screenshots or Code**
```python
from __future__ import annotations

from typing import Generic, TypeVar, TypeVarTuple, Unpack

class Builder(Generic[*Ts]):
    contents: tuple[*Ts]

    def __init__(self, *args: Unpack[Ts]) -> None:
        self.contents = args

    def prepend(self, arg: T) -> Builder[T, *Ts]:
        return Builder(arg, *self.contents)

    def append(self, arg: T) -> Builder[*Ts, T]:
        return Builder(*self.contents, arg)
```
This code is reported as incorrect by pyright, with errors:
```console
  /Users/ellie/Code/python-stuff/python_stuff/variadic.py:16:30 - error: Argument of type "*Ts@Builder" cannot be assigned to parameter "args" of type "*Ts@Builder" in function "__init__"
    Type "*tuple[T@prepend]" cannot be assigned to type "*Ts@Builder" (reportGeneralTypeIssues)
  /Users/ellie/Code/python-stuff/python_stuff/variadic.py:19:40 - error: Expected 1 positional argument (reportGeneralTypeIssues)
```

However when writing the method definition like so:

```python
class Builder(Generic[*Ts]):
    contents: tuple[*Ts]

    def __init__(self, *args: Unpack[Ts]) -> None:
        self.contents = args

    def prepend(self, arg: T) -> Builder[T, *Ts]:
        return Builder(*(arg, *self.contents))

    def append(self, arg: T) -> Builder[*Ts, T]:
        return Builder(*(*self.contents, arg))
```

**VS Code extension or command-line**
```console
$ pyright --version
pyright 1.1.278
```

**Comment by erictraut** [comment]
Yes, this is intended behavior. The argument/parameter matching logic in pyright is already very complicated, and support for variadic TypeVarTuples makes it all the more so. When matching arguments to a TypeVarTuple parameter, pyright requires that the set of arguments comes from either an unpacked argument or a set of regular arguments, not a combination of the two.

The workaround I recommend is the one you've already discovered — build a tuple and unpack it.
